### ISSUE TITLE
[java] `CommentDefaultAccessModifier` false-positive with `static void main` on Java 25

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#6117]
**Affects PMD Version:** 7.17.0

**Rule:** CommentDefaultAccessModifier

Please provide the rule name and a link to the rule documentation:
https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#commentdefaultaccessmodifier

**Description:**
Java 25 allows `public` to be skipped. So this rule shouldn't mark a violation on `static void main`.

**Code Sample demonstrating the issue:**

```java
class Foo {
  static void main(String[] args) {
  }
}
```

**Expected outcome:**
PMD reports a violation at line 2, but that's wrong. That's a false positive.

**Running PMD through:** *Gradle*

**Comment by oowekyala** [comment]
Should we really whitelist main methods? I think this rule is meant to enforce a consistent code style, and making exceptions for it make it less consistent... Are there already exceptions to the rule? 

Edit: one has to consider that main methods are not common in the grand scheme of things, so making an exception for this will look odd I think.

**Comment by zbynek** [comment]
> Are there already exceptions to the rule

There's the exception for all Junit5 test methods (implemented through annotations).

Maybe instead of focusing on main methods we could consider
1) making an exception for compact classes -- those cannot be referenced from the outside, so the concept of method visibility is (almost?) irrelevant there
2) qdding an option to allow default modifier for methods in default classes -- in those I would argue that using `public` modifier is more confusing than the default (if the class has no public subclasses, the public methods are effectively package-private).

**Comment by adangel** [comment]
There is already an exception implemented - it's actually just using the rule as intended. The solution is simple:

```java
class Foo {
  /* default */ static void main(String[] args) {
  }
}
```

Other solutions would be to suppress it for the main class or don't use the rule at all if you don't want to comment default access modifiers.

I agree with @oowekyala , that we shouldn't add more exceptions to this rule.

> Edit: one has to consider that main methods are not common in the grand scheme of things, so making an exception for this will look odd I think.

Exactly. I don't think, this is a false positive at all. The rule works as expected.

**Comment by boris-petrov** [comment]
@adangel the problem with your solution is that the comment says `default` - but that's not true in this case. Not putting a visibility modifier to that method makes it PUBLIC, not package-private. Not the biggest deal though, and I completely agree with you that it's unnecessary to complicate things.

**Comment by adangel** [comment]
It is only public for launching (via java). The main method otherwise behaves as usual - it is package private and can only be access within the same package (or in a subclass).

```
$ cat pkg1/Test.java 
package pkg1;

public class Test {
  void main() {
    System.out.println("Hello");
  }
}
$ cat pkg2/Test2.java 
package pkg2;

class Test2 {
  void main() {
    System.out.println("Test2");
    new pkg1.Test().main();
  }
}

# Running pkg1/Test works:
$ java pkg1/Test.java
Hello

# Running pkg2/Test2 doesn't work:
$ java pkg2/Test2.java
pkg2/Test2.java:6: error: main() is not public in Test; cannot be accessed from outside package
    new pkg1.Test().main();
                   ^
1 error
error: compilation failed
```

**Comment by rkrisztian** [comment]
I think there is a misunderstanding here. Nowadays, IntelliJ IDEA will show the `public` keyword in gray color and hint about it in code like this:

```java
@SpringBootApplication
public class MyApplication {

	public static void main(String[] args) {
		SpringApplication.run(MyApplication.class, args);
	}

}
```

When I hover on the `public` keyword, IntelliJ IDEA says, *Modifier 'public' is redundant for 'main' method on Java 25*, and offer that I remove that modifier. When I do, the Spring application will still start fine, but PMD thinks this is a mistake:

> [Missing commented default access modifier on method 'main(String[])'](https://docs.pmd-code.org/pmd-doc-7.13.0/pmd_rules_java_codestyle.html#commentdefaultaccessmodifier)

I could add `/* default */`, but what good does it do? Practically we gained nothing from Java 25, because now we got something more verbose to type. When I type `psvma` and press Enter, IDEA will insert:

```java
	static void main(String[] args) {
		
	}
```

No public modifier, again. Adding the `/* default */` comment doesn't feel right, because it's not pragmatic to type.

Not only that, without the `public` modifier, PMD will also report:

> [Consider using varargs for methods or constructors which take an array the last parameter.](https://docs.pmd-code.org/pmd-doc-7.13.0/pmd_rules_java_bestpractices.html#usevarargs)

This seems also inconsistent. I assume that with the public modifier, PMD does not report a violation because `String[] args` is the traditional and more widely recognized form for the `main` method parameters. However, without the modifier, this is no longer recognized.
